src/mame/snk/k1ge.cpp

```diff
@@ -59,77 +61,130 @@ uint8_t k1ge_device::read(offs_t offset)
+	case 0x100: case 0x101: case 0x102: case 0x103:
+	case 0x104: case 0x105: case 0x106: case 0x107:
+	case 0x108: case 0x109: case 0x10a: case 0x10b:
+	case 0x10c: case 0x10d: case 0x10e: case 0x10f:
+	case 0x110: case 0x111: case 0x112: case 0x113:
+	case 0x114: case 0x115: case 0x116: case 0x117:
 		data &= 0x07;
+		set_pen_indirect(offset - 0x100, data);
+		break;
+	case 0x118:
+		set_pen_indirect(BG_COLOR, ((data & 0xc0) == 0x80) ? data & 7 : 0);
 		break;
 	case 0x7e2:
-		if ( m_vram[0x7f0] != 0xaa )
+		if (m_vram[0x7f0] != 0xaa)
 			return;
 		data &= 0x80;
 		break;
 	}
 
+	m_vram[offset] = data;
+}
+
+
+void k2ge_device::write(offs_t offset, u8 data)
+{
+	assert(offset < 0x4000);
+
+	switch (offset)
+	{
+	case 0x100: case 0x101: case 0x102: case 0x103:
+		data &= 0x07;
+		set_pen_indirect(offset - 0x100 + MONO_COLOR, MONO_COLOR + data);
+		break;
+	case 0x104: case 0x105: case 0x106: case 0x107:
+		data &= 0x07;
+		set_pen_indirect(offset - 0x100 + MONO_COLOR, MONO_COLOR + 0x08 + data);
+		break;
+	case 0x108: case 0x109: case 0x10a: case 0x10b:
+		data &= 0x07;
+		set_pen_indirect(offset - 0x100 + MONO_COLOR, MONO_COLOR + 0x10 + data);
+		break;
+	case 0x10c: case 0x10d: case 0x10e: case 0x10f:
+		data &= 0x07;
+		set_pen_indirect(offset - 0x100 + MONO_COLOR, MONO_COLOR + 0x18 + data);
+		break;
+	case 0x110: case 0x111: case 0x112: case 0x113:
+		data &= 0x07;
+		set_pen_indirect(offset - 0x100 + MONO_COLOR, MONO_COLOR + 0x20 + data);
+		break;
+	case 0x114: case 0x115: case 0x116: case 0x117:
+		data &= 0x07;
+		set_pen_indirect(offset - 0x100 + MONO_COLOR, MONO_COLOR + 0x28 + data);
+		break;
+	case 0x118:
+		set_pen_indirect(BG_COLOR, MONO_COLOR + 0x30 + (((data & 0xc0) == 0x80) ? (data & 7) : 0));
+		break;
+	default:
+		k1ge_device::write(offset, data);
+		break;
+	}
+
 	/* Only the lower 4 bits of the palette entry high bytes can be written */
-	if ( offset >= 0x0200 && offset < 0x0400 && ( offset & 1 ) )
+	if (offset >= 0x0200 && offset < 0x0400 && BIT(offset, 0))
 	{
 		data &= 0x0f;
 	}
 
 	m_vram[offset] = data;
+
+	// set palette
+	if (offset >= 0x0200 && offset < 0x0400)
+	{
+		u16 const palette = m_vram[offset & ~1] | (m_vram[offset | 1] << 8);
+		u8 const r = pal4bit(palette & 0xf);
+		u8 const g = pal4bit((palette >> 4) & 0xf);
+		u8 const b = pal4bit((palette >> 8) & 0xf);
+		set_indirect_color((offset - 0x0200) >> 1, rgb_t(r, g, b));
+	}
 }
 
 
-void k1ge_device::draw_scroll_plane( uint16_t *p, uint16_t base, int line, int scroll_x, int scroll_y, int pal_base )
+void k1ge_device::draw_scroll_plane(u16 *p, u16 base, int line, int scroll_x, int scroll_y, int pal_base)
 {
-	int i;
-	int offset_x = ( scroll_x >> 3 ) * 2;
+	int offset_x = (scroll_x >> 3) * 2;
 	int px = scroll_x & 0x07;
-	uint16_t map_data;
-	uint16_t hflip;
-	uint16_t pcode;
-	uint16_t tile_addr;
-	uint16_t tile_data;
 
-	base += ( ( ( ( scroll_y + line ) >> 3 ) * 0x0040 ) & 0x7ff );
+	base += ((((scroll_y + line) >> 3) * 0x0040) & 0x7ff);
```

Comment:
The 0x7ff mask doesn’t really represent the actual values you can get because the left shift means the low six bits will be zero.  This would be clearer if the mask happened before the scaling:
```c++
	base += (((scroll_y + line) >> 3) & 0x1f) * 0x0040;
```